File: PerlInterviewQuestions/numberOfPaths.pl
use strict;
use warnings;
use Data::Dumper;

sub displayList {
    my $L = shift();

    for (my $i = 0; $i < scalar(@$L); $i++) {
        for (my $j = 0; $j < scalar(@{$$L[$i]}); $j++) {
            print $$L[$i][$j] . " ";
        }
        print "\n";
    }
}

sub generateMatrix {
    my $n = shift;
    my $length = shift;
    my @L = ();
    for (my $i = 0; $i < $length; $i++) {
        my @A = ();
        for (my $j = 0; $j < $n; $j++) {
            push(@A, 0);
        }
        push(@L, \@A);
    }
    return @L;
}

sub numberOfPaths {
    my $n = shift;
    my $length = shift;
    my @L = generateMatrix($n, $length);
    for (my $i = 0; $i < scalar(@L); $i++) {
        for (my $j = 0; $j < scalar(@{$L[$i]}); $j++) {
            if (($i == 0 && $j != 0) || ($i != 0 && $j == 0)) {
                $L[$i][$j] = 1;
            }

            if ($i != 0 && $j != 0) {
                $L[$i][$j] = $L[$i - 1][$j] + $L[$i][$j - 1];
            }
        }
    }
    return @L;
}

my @L = numberOfPaths(4, 5);
displayList(\@L);

File: PythonInterviewQuestions/numberOfPaths.py
def genarateZeroMatrix(n, m):
    A = [[0 for x in range(n)] for x in range(m)]
    return A


def displayList(a):
    for i in range(len(a)):
        for j in range(len(a[i])):
            print(a[i][j], end=" ")
        print()


def numberOfPaths(n, m):
    A = genarateZeroMatrix(n, m)
    for i in range(m):
        for j in range(n):
            if not (i == 0 and j == 0):
                if (i == 0 and j != 0) or (i != 0 and j == 0):
                    A[i][j] = 1
                else:
                    A[i][j] = A[i - 1][j] + A[i][j - 1]

    return A


def main():
    L = numberOfPaths(4, 5)
    displayList(L)


if __name__ == '__main__':
    main()
